lsp/server/src/diagnostic/DiagnosticProducer.ts

```diff
@@ -7,16 +7,28 @@
 
 import { Diagnostic } from 'vscode-languageserver/node';
 import { TextDocument } from 'vscode-languageserver-textdocument';
+import { ProducerId } from './DiagnosticSettings';
 
 export interface DiagnosticProducer<T> {
+
+    /**
+     * Get the Id for the diagnostic producer. 
+     */
+    getId(): ProducerId;
```

Comment:
Is string type enough?

Reply:
@haifeng-li-at-salesforce @sfdctaka I over thought about the enforcing producer id uniqueness by defining the in cetral place. let me remove the ProducerId thing.  hope the individual producer will pick the same id.

Reply:
This point that you make, is what I wanted to hear in my other question about the benefit of having union type of string in `ProducerId`. So you want a unique ID for producer.

I still think union type is not the way to go.

I think @haifeng-li-at-salesforce has a great point. Produce ID probably doesn't suffice what's trying to be solved here. The problem is to be able to filter/suppress diagnostics. And to effectively do that in runtime without name collision is to have a unique compile time ID and to use that ID instead of names(string). Simple thinking me says give a static GUID to the diagnostic rule. You can make this a `ProducerId` if you like. And you'll add another property in `Diagnostics` that's used only for displaying name of that diagnostic.

Reply:
removed.